### PR TITLE
slcli vlan cancel should report if a vlan is automatic

### PR DESCRIPTION
**Pull request by caberos**
https://github.com/softlayer/softlayer-python/issues/1489

### CODE REVIEW

File: SoftLayer/CLI/vlan/cancel.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+@click.argument('identifier')
+@environment.pass_env
+def cli(env, identifier):
+    """Cancel network vlan."""
+
+    mgr = SoftLayer.NetworkManager(env.client)
+    billing = BillingManager(env.client)
+    if not (env.skip_confirmations or formatting.no_going_back(identifier)):
+        raise exceptions.CLIAbort('Aborted')
+
```
Comment (ATGE) [inline SoftLayer/CLI/vlan/cancel.py:23]: adding a resolve vlan name as `helpers.resolve_id(manager.resolve_ids, volume_id, 'Network Vlan')` would be nice

File: SoftLayer/managers/billing.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+"""
```
Comment (ATGE) [inline SoftLayer/managers/billing.py:1]: As the mostly managed Manager modules their own billing stuff, maybe it's better to move this logic into the Network Manager, but in any case, keep in mind to add tests for that.

**Review by FernandoOjeda — COMMENTED** [review]
Do not forget to add unit test for the new methods implemented.

File: SoftLayer/CLI/vlan/cancel.py (hunk 1 above)
Comment (allmightyspiff) [reply SoftLayer/CLI/vlan/cancel.py:23]: The problem with a resolver here is that vlan Number and vlan Name are both not unique fields, so you might accidently cancel a vlan you didn't mean to. Lets just stick with vlan ID here.

File: SoftLayer/CLI/vlan/cancel.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    item = mgr.get_vlan(identifier).get('billingItem')
+    if item:
+        billing.cancel_item(item.get('id'), 'cancel by cli command')
```
[the remark this answers is not in the document]
Comment (allmightyspiff) [reply SoftLayer/CLI/vlan/cancel.py:26]: Yes, we can let the customer add a reason, but having the default be "Cancelled from SLCLI" is good to. I know that I would hate to have to put a cancellation reason everytime, and I think most customers don't care what the reason is.

File: SoftLayer/managers/network.py [hunk 4]
```diff
@@ -752,3 +753,10 @@ def set_subnet_ipddress_note(self, identifier, note):
         """
         result = self.client.call('SoftLayer_Network_Subnet_IpAddress', 'editObject', note, id=identifier)
         return result
+
+    def get_cancel_failure_reasons(self, identifier):
+        """get the reasons by cannot cancel the VLAN
```
Comment (allmightyspiff) [inline SoftLayer/managers/network.py:758]:
```suggestion
        """get the reasons why we cannot cancel the VLAN
```

**Review by allmightyspiff — CHANGES_REQUESTED** [review]
Change the logic of CLI/vlan/cancel.py to something like this:



```
reasons = mgr.get_cancel_failure_reasons(identifier)
if len(reasons) > 0:
    raise exceptions.CliAbort(reasons )
item = mgr.get_vlan(identifier).get('billingItem')
if item:
    billing.cancel_item(item.get('id'), 'cancel by cli command')
else:
    raise exceptions.CliAbort("VLAN is an automatically assigned and free of charge VLAN, it will automatically be removed from your account when it is empty")

```

**Review by allmightyspiff — APPROVED** [review]
Changes look good, thanks.

**Review by FernandoOjeda — APPROVED** [review]
Looks good.

File: SoftLayer/CLI/vlan/cancel.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from SoftLayer.CLI import environment
+from SoftLayer.CLI import exceptions
+from SoftLayer.CLI import formatting
+
+
+@click.command()
+@click.argument('identifier')
+@environment.pass_env
+def cli(env, identifier):
+    """Cancel network vlan."""
```
Comment (ATGE) [inline SoftLayer/CLI/vlan/cancel.py:16]:
```suggestion
    """Cancel a Network VLAN."""
```

**Review by ATGE — APPROVED** [review]
### Testing
- ✔️slcli  vlan cancel 1234556
- ✔️slcli -v -y vlan cancel 1234556
- ✔️slcli vlan cancel -h' 
-✔️ slcli vlan -h

looks good, just a comment on cancel cli method docs.

**Pull request merged**